src/range_util.rs

```diff
@@ -0,0 +1,532 @@
+        let mut padded = BytesMut::from(b.as_ref());
+        while padded.len() < len {
+            padded.put_u8(value);
+        }
+        padded.freeze()
+    }
+
+    fn inclusive_end_bound(range: &BytesRange, len: usize) -> Bytes {
+        match range.end_bound() {
+            Unbounded => padded_bytes(&Bytes::new(), u8::MAX, len),
+            Included(b) | Excluded(b) => {
+                let b = padded_bytes(&b, u8::MIN, len);
+                decrement_lex(&b)
+            }
+        }
+    }
+
+    fn inclusive_start_bound(range: &BytesRange, len: usize) -> Bytes {
+        match range.start_bound() {
+            Unbounded => padded_bytes(&Bytes::new(), u8::MIN, len),
+            Included(b) => padded_bytes(&b, u8::MIN, len),
+            Excluded(b) => {
+                let b = padded_bytes(&b, u8::MIN, len);
+                increment_lex(&b)
+            }
+        }
+    }
+
+    fn can_decrement_without_truncation(bytes: &Bytes) -> bool {
+        bytes.iter().any(|b| *b > u8::MIN)
+    }
+
+    fn arbitrary_minvalue_bytes(min_bound: Bound<usize>, max_bound: Bound<usize>) -> Bytes {
+        let min_len = match min_bound {
+            Unbounded => 0,
+            Included(len) => len,
+            Excluded(len) => len + 1,
+        };
+
+        let max_len = match max_bound {
+            Unbounded => min_len + 10,
+            Included(len) => len + 1,
+            Excluded(len) => len,
+        };
+
+        let len = if min_len == max_len {
+            min_len
+        } else {
+            min_len + (random::<usize>() % (max_len - min_len))
+        };
+
+        Bytes::from(vec![u8::MIN; len])
+    }
+
+    pub(crate) fn arbitrary_bytes_in_range(range: &BytesRange) -> Bytes {
+        assert!(
+            range.non_empty(),
+            "Cannot choose arbitrary value from an empty range"
+        );
+
+        if let Some(end) = range.end_bound_opt() {
+            if !can_decrement_without_truncation(&end) {
+                let min_len = range.start_bound().map(|b| b.len());
+                let max_len = range.end_bound().map(|b| b.len());
+                return arbitrary_minvalue_bytes(min_len, max_len);
+            } else if range.start_bound() == Included(&end) {
+                return end.clone();
+            } else {
+                let start = range.start_bound_opt().unwrap_or_else(|| Bytes::new());
+                if is_prefix_increment(&start, &end) {
+                    return if range.start_bound() == Included(&start) {
+                        start.clone()
+                    } else {
+                        end.clone()
+                    };
+                }
+            }
+        }
+
+        let len = max(bound_len(range.start_bound()), bound_len(range.end_bound())) + 3;
+
+        let start_bound = inclusive_start_bound(&range, len);
+        let end_bound = inclusive_end_bound(&range, len);
+        arbitrary_bytes_between(&start_bound, &end_bound)
+    }
+
+    proptest! {
```

Comment:
How do you think we should handle failures from these, because as I understand it they aren't deterministic. So this means that the test may fail randomly on a pr branch or main. We probably want to update the CI automation to check in the regressions file so that when a failure is found it starts failing consistently.

Reply:
I think the idea is that the regressions file should get checked in to the source repository. Then proptest automatically runs the tests against it.